arch/risc-v/src/mpfs/mpfs_can.c

```diff
@@ -0,0 +1,2953 @@
+        */
+
+      priv->dev.d_buf = (uint8_t *)priv->txdesc;
+    }
+
+  /* Check for RX FIFO Overflow */
+
+  if (MPFS_CAN_INT_STATUS_OVR_LOAD & priv->isr)
+    {
+      /* Re-enable RX overload err int as all the RX buffers are handled */
+
+      mpfs_can_set_int_ebl(priv, MPFS_CAN_INT_ENABLE_OVR_LOAD_INT_ENBL);
+    }
+
+  /* Re-enable RX msg receive interrupt */
+
+  mpfs_can_set_int_ebl(priv, MPFS_CAN_INT_ENABLE_RX_MSG_INT_ENBL);
+}
+
+/****************************************************************************
+ * Name: mpfs_txdone_work
+ *
+ * Description:
+ *  An interrupt was received indicating that the last TX packet(s) is done
+ *
+ * Input Parameters:
+ *  priv  - Pointer to the private CAN driver state structure
+ *
+ * Returned Value:
+ *  None
+ *
+ * Assumptions:
+ *  Global interrupts are disabled by the watchdog logic.
+ *  We are not in an interrupt context so that we can lock the network.
+ *
+ ****************************************************************************/
+
+static void mpfs_txdone_work(void *arg)
+{
+  mpfs_can_instance_t *priv = (mpfs_can_instance_t *)arg;
+
+#ifdef CONFIG_DEBUG_CAN_INFO
+  caninfo("TX done interrupt received\n");
+#endif
+
+  /* There should be space for a new TX in any event.  Poll the network for
+   * new XMIT data
+   */
+
+  net_lock();
+  devif_poll(&priv->dev, mpfs_txpoll);
+  net_unlock();
+}
+
+/****************************************************************************
+ * Name: mpfs_transmit
+ *
+ * Description:
+ *  Start hardware transmission.  Called either from the txdone interrupt
+ *  handling or from watchdog based polling.
+ *
+ * Input Parameters:
+ *  priv  - Pointer to the private CAN driver state structure
+ *
+ * Returned Value:
+ *  Zero (CAN_OK) on success; a negated errno on failure
+ *
+ * Assumptions:
+ *  May or may not be called from an interrupt handler.  In either case,
+ *  global interrupts are disabled, either explicitly or indirectly through
+ *  interrupt handling logic.
+ *
+ ****************************************************************************/
+
+static int mpfs_transmit(mpfs_can_instance_t *priv)
+{
+  uint8_t ret;
+  uint8_t i;
+
+  /* Retrieve the CAN 2.0B frame from network device buffer */
+
+  struct can_frame *cf = (struct can_frame *)priv->dev.d_buf;
+
+  /* Fill the CAN msg object to be sent */
+
+  pmpfs_can_msgobject_t pmsg = priv->tx_msg;
+
+  /* CAN TX msg control command */
+
+  pmsg->msg_ctrl = ((cf->can_id & CAN_EFF_FLAG) ?
+                   MPFS_CAN_TX_MSG_CTRL_CMD_IDE : 0)
+                   | ((cf->can_dlc << MPFS_CAN_TX_MSG_CTRL_CMD_DLC_SHIFT)
+                   & MPFS_CAN_TX_MSG_CTRL_CMD_DLC)
+                   | ((cf->can_id & CAN_RTR_FLAG) ?
+                   MPFS_CAN_TX_MSG_CTRL_CMD_RTR : 0);
+
+  /* CAN ID */
+
+  pmsg->id = mpfs_can_canid_to_msgid(cf->can_id);
+
+  /* CAN data */
+
+  if (!(pmsg->msg_ctrl & MPFS_CAN_TX_MSG_CTRL_CMD_RTR))
+    {
+      for (i = 0; i < cf->can_dlc; i++)
```

Comment:
This is again unnecessary complicated, imho. you can just copy&bswap both 32-bit words as in receive

Reply:
yes, that's much more elegant operation. i forgot about that. thanks!